Authorization middleware

peteranny opened the pull request:
add middleware/authorization.js

```
function(request, response, next){
  if(found_in_redis){
    resolve()
  } else {
    if(data_num + ref_num>0){
      resolve();
      inert_in_redis();
    } else {
      reject();
    }
  }
  then
  if(resolve) next();
  if(reject) throw 403;
}
```

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,107 @@
+            reject(err);
+        });
+    });
+}
```
`MongoDB` 的 API 返回已經是 Promise 物件了，不需要重複包，我把你的 code 的外匡移除如下

```
function getDataNumOfUser(user_id, db) {
  return db.collection('authors').find(xxx).toArray().then(results => {
    if (results.length == 0) {
      return 0;
    } else {
      return results[0].queries_count || 0; // we the queries_count field is missing
    }
  })
}
```

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,107 @@
+        if (sum > 0) {
+            return Promise.resolve();
+        } else {
+            return Promise.reject("User does not meet authorization level");
```
No permission 應該要會傳 return false （你的函數名是有無權限）；有權限則是 true，用reject會讓接的 err 分不出是 db error 或是其他可能

或是你要重新命名你的函數名稱，讓它 make sense

BTW, 看起來是要用 `const sum`

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,107 @@
+    then(
+    () => {
+        return Promise.resolve();
+    }, err => {
```
in this case, you can write `redisLookUp(xxx).catch(err => {xxx})`

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,107 @@
+        return hasSearchPermission(request.user_id, request.db)
+        // write authorized user into cache for later access
+        .then(() => {
+            return redisInsert(request.user_id, request.redis_client);
```
這邊要小心處理：redisInsert 失敗（rejected）不代表 user 沒有權限，還是OK的，所以當 redisInsert 是 reject 的話，應該要捕捉起來，而不是將這個 reject reason 交到 L105 處理

mark86092 commented:
你會寫一個小測試測試它嗎？

peteranny commented:
我知道要寫`test/testAuthorization.js`
但........ 不會寫T T

我只知道`middleware`要把redis先初始化好

```
const middleware = require('../middlewares').expressRedisDb('');
middleware(req, {}, function() {
    ....
});
```

然後`before`要先把測資放進db
但其它怎麼寫就不會了

```
before(function() {
  return MongoClient
    .connect(process.env.MONGODB_URI)
    .then(function(_db) {
        db = _db;
        const test_data = [
            {id: 'a', queried_count: 1, reference_count: 1},
            {id: 'b', queried_count: 1, reference_count: 0},
            {id: 'c', queried_count: 1, reference_count: undefined},
            {id: 'd', queried_count: 0, reference_count: 1},
            {id: 'e', queried_count: 0, reference_count: 0},
            {id: 'f', queried_count: 0, reference_count: undefined},
            {id: 'g', queried_count: undefined, reference_count: 1},
            {id: 'h', queried_count: undefined, reference_count: 0},
            {id: 'i', queried_count: undefined, reference_count: undefined},
        ];
        return Promise.all([
            db.collection('authors').insertMany(
               test_data
              .filter(data => data.queried_count!==undefined)
              .map(data => ({
                 _id: {
                    id: data.id,
                    type: 'facebook',
                },
                queried_count: data.queried_count,
            }))
        ),
       db.collection('references').insertMany(
            test_data
            .filter(data => data.reference_count!==undefined)
            .map(data => ({
                user: {
                    id: data.id,
                    type: 'facebook',
                },
                count: data.reference_count,
            }))
        )
    ]);
})
```

peteranny commented:
Updated and fixed.

`getDataNumOfUser()` and `getRefNumOfuser()` now resolves 0 on undefined values
Fix logic like: both redis insertion resolution and rejection finally resolves search permission

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,102 @@
+        // write authorized user into cache for later access
+        .then(hasSearchPermission => {
+            if (hasSearchPermission) {
+                return redisInsert(request.user_id, request.redis_client).finally(Promise.resolve);
```
Standard Promise 沒有 finally 的實作喔，
```
return redisInsert(request.user_id, request.redis_client).catch(err => {});
```

即可

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,102 @@
+    })
+    // proceed or throw error
+    .then(hasSearchPermission => {
+        next(hasSearchPermission);
```
Expressjs 的 next 有參數代表 pass 一個錯誤過去，所以 L98 這邊應該是 next()

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,102 @@
+    .then(hasSearchPermission => {
+        next(hasSearchPermission);
+    }, err => {
+        throw new HttpError(403, err);
```
所以，這邊應該是 `next(new HttpError("Forbidden", 403))` (請參考 errors 那邊的註解)

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,102 @@
+            return Promise.resolve(false);
+        }
+    }, err => {
+        return Promise.reject(err);
```
這邊可以不需要處理 err, 然後再把 reject err 出去

mark86092 commented on middlewares/authorization.js:
```diff
@@ -0,0 +1,102 @@
+    ])
+    // determine authorization
+    .then(values => {
+        let sum = values.reduce((a, b) => a+b);
```
這邊的 sum 是 `const sum`

peteranny commented:
`module.exports` should return a `promise` -- fixed

Add `test/testAutorization.js`

mark86092 commented:
> module.exports should return a promise

Why it should? It is a middleware

peteranny commented:
我昨天研究才發現，promise chain會無法throw error （會被promise吸收成rejection）
昨天弄好久實在不知道怎麼做，還是 @mark86092 教我一下怎麼在promise chain最後throw error ><

```
module.exports = (request, response, next) => {
    ...
    .then(() => {
        next();
    }, err => {
        throw new HttpError('Forbidden', 403);
        // ^^^ This throw will be absorbed as Promise.reject()
    });
};
```

mark86092 commented:
在 middleware 中，我們需要的是把 error next 到下一層去，

所以，當 promise 處於 rejected 時，只要 `next(new HttpError("Forbidden", 403))` 就可以了

mark86092 commented:
或許你可以參考看看 https://github.com/goodjoblife/WorkTimeSurvey-backend/pull/140/files#diff-95135aca7be408a65339d7df176c1508 裡面的寫法

peteranny commented:
Done. Have a look.

mark86092 commented:
看起來有測試到 queries_count + reference_count 滿足後會通過 middleware，
不過哪一段有測試到 如果 redis cached，就 bypass mongodb 的 query ?

mark86092 commented on test/testAuthorization.js:
```diff
@@ -0,0 +1,89 @@
+            after(function() {
+                req.db.collection('authors').remove({});
+                req.db.collection('references').remove({});
+                req.redis_client.flushall();
```
L82~L84 是 async 指令，要確保它們做完

peteranny commented:
Checkout redis這個會檢查redis有沒有資料，但我不知道怎麼確認pass是redis有而通過，還是redis沒有再檢查才通過。

mark86092 commented:
了解，這個我再補，

BTW, 想麻煩你把這三個函數 module.exports 出來，

getDataNumOfUser
getRefNumOfUser
resolveSearchPermission

peteranny commented:
想知道export出來的意義是？這是常用功能嗎@@

（我還是會export啦）

mark86092 commented:
幾個原因：
1. 那三個的確是未來有機會會用到
2. authorization 現在雖然只有一種，但也許會擴充
3. 在測試上面可能會用到

peteranny commented:
Done.

mark86092 commented on test/testAuthorization.js:
```diff
@@ -0,0 +1,95 @@
+        })
+        // setup redis
+        .then(new Promise(function(resolve, reject) {
+            require('../middlewares').expressRedisDb('')(req, {}, err => {
```
剛剛自己docker環境測試發現不過，原來是 `expressRedisDb('')` --> `expressRedisDb(process.env.REDIS_URL)`，我補測試時再改

peteranny commented on test/testAuthorization.js:
```diff
@@ -0,0 +1,95 @@
+            });
+
+            it('checkout redis', function(done) {
+                // second access to permission will checkout redis first
```
原本是想說，第一次call authorization後，直接再call一次authorization，這樣子第二次理論上就是會先讀取redis，不過因為其實我也無從判斷authorization的pass是redis得到還是沒得到再算一次才得到，所以想說乾脆在第一次call之後，直接簡查redis有沒有（就是這個checkout redis）

不過我是有用console.log檢查過，第二次call都是從redis拿沒有錯，如果第一次沒有pass第二次就會redis拿不到而重算，這個也沒錯
